### ISSUE TITLE
[Feature request] Negative padding for F.pad but not convolution functions?

### ISSUE THREAD

**Issue by jrdurrant** [pytorch/pytorch#4939]
I am using a Conv2d layer that requires negative padding, and I can achieve this by using F.pad and Conv2d with appropriate padding values. It seems like it would be cleaner and possibly more efficient(?) to be able to do it with only the layer parameters though. 

Is there a reason why the valid padding values for convolution functions are not consistent with F.pad?

**Comment by soumith** [comment]
the padding in Conv2d layer is implemented as implicit padding, i.e. the convolution kernel itself is assuming that the given input is padded and doing the computation. There is no extra memory taken by the operation because of the padding value.

The F.pad layer does padding more explicitly, i.e. each input is padded and an output Tensor is returned. This involves extra memory.

**Comment by jrdurrant** [comment]
Ah, yes I understand. Thanks for such a fast response.

For negative padding does this mean it should not be an option for the Conv2d layer (since it breaks the assumption that padding is explicitly adding zeros) or it would make sense to implement it as an option (given that there's no extra computation, but instead adjusting pointers to implicitly have a smaller image)?

**Comment by fmassa** [comment]
I think supporting negative padding in the operations shoudn't require too many changes to the code, but I'm curious: what's the use-case for negative padding (which means cropping the input)?
We usually want to keep as much information as possible of the inputs, and a negative padding would throw away such information.

**Comment by jrdurrant** [comment]
@fmassa Yes exactly, and that certainly explains why I can't seem to find any information elsewhere either! I have an implementation of the paper [Convolution Pyramids](http://www.cs.huji.ac.il/labs/cglab/projects/convpyr/) - where their 'pyramid' is effectively an encoder-decoder network with skip connections - that I think this makes most sense for. 

In order for their method to work correctly the convolutions must be 'full' convolutions in the encoder part of the network. This means the image is being padded such that it actually ends up larger (before being downsampled) at each layer. In order for the sizes of the equivalent layers with skip connections to work this means doing the inverse of this during the decoder part - this is the negative padding. So although the edges of the image are being discarded, there is no actual _information_ being lost.

**Comment by fmassa** [comment]
I think I see what you mean.
I believe in those cases one would use a combination of `Conv` and `ConvTranspose` (also known as `FullConv` or `Deconv`), and `ConvTranspose` [has a parameter `output_size`](http://pytorch.org/docs/0.3.0/nn.html#torch.nn.functional.conv_transpose2d) which accounts for some ambiguity in the output size, which could be seen as some sort of "cropping".

Is that enough for your use-case?

**Comment by jrdurrant** [comment]
I can't find a parameter `output_size` there? Making use of `ConvTranspose` for this seems like a really good idea though. I have `ConvTranspose` layers at the moment, but I was only using them for the 'zero' upsampling. From a quick test I think you're right, it seems as though I can do what I need in terms of padding there as well. Thanks for your help

**Comment by fmassa** [comment]
@jrdurrant I meant `output_padding`, I mixed that with `MaxUnpooling`, sorry :)
Can this issue be closed then?

**Comment by jrdurrant** [comment]
@fmassa No worries at all, that was exactly what I needed, thanks. And yes, it seems as though this is unusual enough that it doesn't justify add support for it in general

**Comment by ZisIsNotZis** [comment]
But implementing negative-padded `Conv` using `ConvTranspose` only works when `stride=1` (because you can't set `stride=0.5` in `ConvTranspose`). If I want to do a 1x1 convolution with stride 2, but at all *even* positions, like

```
.....
.x.x.
.....
.x.x.
.....
```

instead of

```
x.x.x
.....
x.x.x
.....
x.x.x
```

I have to use a `ZeroPad2d` with `-1` pad first. This evenness is needed to match the shape of a kernel 3 stride 2 not-padded convolution (so that they can added up).

Therefore I believe negative padding in `Conv` is meaningful because it removes the explicit padding process in `ZeroPad2d`? Sorry if I got it wrong (e.g. maybe ZeroPad2d with negative pad takes no time because it uses stride tricks or etc).

**Comment by SvenPVoigt** [comment]
Negative padding still isn't implemented in pytorch 1.3.1. It would be useful to help match the output size of 1x1 convolutions with 3x3 convolutions when doing concatenation. My thought is that padding requires an extra (2*(m-2) + 2*(n-4)) 3x3 convolutions. However, by cropping the edges we wouldn't need those extra 3x3 convolutions and also need less 1x1 convolutions (bc we would negative pad the 1x1 convolutions).
